### ISSUE TITLE
Scheduled transactions don't handle 29th/30th/31st dates appropriately

### ISSUE THREAD

**Issue by EKCJ** [actualbudget/actual#1062]
Running v23.5.0, when I set up a monthly scheduled transaction starting 31/5 the current behaviour is to not schedule anything for June, because there is no 31st day in June.
![image](https://github.com/actualbudget/docs/assets/67192679/7e479f69-838d-475f-9df7-5c060182412a)
I'm sure opinions will differ on whether the best option would be to schedule the transaction for 30/6 or for 1/7, I don't have an opinion on that as long as it gets scheduled somewhere. 🙂

**Comment by Kidglove57** [comment]
I have always used the last day of the month option. In my experience, some payees will take it then and others on the 1st. 

Only my view but I think the current option is adequate and probably the only reasonable solution. 

I believe there is another issue live for the more thorny problem of payments due on a weekend?

**Comment by EKCJ** [comment]
Many thanks for moving this @j-f1 

The last day of month option is useful and does solve the problem for the 31st - however there's still an issue with payments which should always be on the 29th or 30th, even if the month is longer, and only move forward if the month is shorter.

I haven't tried a weekend yet so won't comment on that topic 😅

**Comment by rich-howell** [comment]
> Many thanks for moving this @j-f1
> 
> The last day of month option is useful and does solve the problem for the 31st - however there's still an issue with payments which should always be on the 29th or 30th, even if the month is longer, and only move forward if the month is shorter.
> 
> I haven't tried a weekend yet so won't comment on that topic 😅

An example would be Feb. If you have something going out on the 30th in January Feb would get ignored and March would pick up again. 

The end of the month wouldn't work in the above situation as the 30th isn't the end of the month all the time as @EKCJ pointed out. 

This seems like a bug.

**Comment by Shazib** [comment]
The [package](https://gitlab.com/john.carroll.p/rschedule) that Actual uses to create the schedules, appears to implement the ICAL spec  [RFC5545](https://datatracker.ietf.org/doc/html/rfc5545) for date recurrences, 

From some cursory googling, I dont think the spec can actually cope with this

**Comment by lukepetrolekas** [comment]
No, an internal system does not need to be created. The ICalendar spec can absolutely handle the case of "last day of month" and "last workday of month" scenarios.

Also [From the ICalendar spec definition of Recurrence Relation](https://icalendar.org/iCalendar-RFC-5545/3-3-10-recurrence-rule.html)

> The BYSETPOS rule part specifies a COMMA-separated list of values that corresponds to the nth occurrence within the set of recurrence instances specified by the rule. BYSETPOS operates on a set of recurrence instances in one interval of the recurrence rule. For example, in a WEEKLY rule, the interval would be one week A set of recurrence instances starts at the beginning of the interval defined by the FREQ rule part. Valid values are 1 to 366 or -366 to -1. It MUST only be used in conjunction with another BYxxx rule part. For example "the last work day of the month" could be represented as:

` FREQ=MONTHLY;BYDAY=MO,TU,WE,TH,FR;BYSETPOS=-1`

I would suspect the last day of the month could then be specified as:

` FREQ=MONTHLY;BYDAY=MO,TU,WE,TH,FR,SA,SU;BYSETPOS=-1`

So long as the rest of the RRULE does not specifically select a date such as the 29, 30, or 31st, I think it should work. It should be implemented as a radio button or some form of either/or scenario.

**Comment by Shazib** [comment]
As I understood it, selecting a specific date is exactly what is wanted? While having it cope with out of bounds dates for certain months

**Comment by lukepetrolekas** [comment]
Ok, I put my foot in my mouth back there, you're correct. I did manage to find this, which looks to be the exact same question.

https://stackoverflow.com/questions/35757778/rrule-for-repeating-monthly-on-the-31st-or-closest-day

It's supported by a newer addition called RSCALE, but not all RRULE libraries support it yet.

**Comment by passabilities** [comment]
where did we land on this? seems to still be an issue as I'm trying to create a schedule for the 30th and it is not automatically picking up February

**Comment by salomonsters** [comment]
This issue occurred for me as well. A recurring payment which is slightly regular (company puts it at 29th or before the weekend) so I can't use end of the month for that - it can be as early as the 27th. So in February my rule didn't trigger.
I worked around it by changing it to 28, manually selecting the transaction and applying the rule, then reverting it back to the 29th. But quite convenient and definitely a bug imho that I would love to see fixed.

**Comment by RMcGhee** [comment]
@passabilities @salomonsters 
There's a PR that would fix this: https://github.com/actualbudget/actual/pull/7156
But I think it's very much worth a discussion about whether it's the right path forward.
